lib/libzfs_core/libzfs_core.c

```diff
@@ -645,6 +656,90 @@ lzc_send_resume(const char *snapname, const char *from, int fd,
 	    resumeoff, NULL));
 }
 
+unsigned long
+lzc_get_pipe_max()
+{
+	unsigned long max_psize = 65536;
+#ifdef __linux__
+	FILE *procf = fopen("/proc/sys/fs/pipe-max-size", "re");
+
+	if (procf != NULL) {
+		if (fscanf(procf, "%lu", &max_psize) <= 0) {
+			// safe default
+			max_psize = 65536;
+		}
+		fclose(procf);
+	}
+#endif
+	return (max_psize);
+}
+
+void
+lzc_set_pipe_max(int infd)
+{
+#ifdef __linux__
+	unsigned long max_psize = lzc_get_pipe_max();
+	long cur_psize;
+	cur_psize = fcntl(infd, F_GETPIPE_SZ);
+	if (cur_psize > 0 &&
+	    max_psize > (unsigned long) cur_psize)
+		fcntl(infd, F_SETPIPE_SZ,
+		    max_psize);
+#endif
+}
```

Comment:
It's awkward to split the `lzc_get_pipe_max` and `lzc_set_pipe_max` functionality across libraries.  I don't see any reason we couldn't move the `libzfs_set_pipe_max()` call from `zfs_receive()` in to the `recv_impll()` function in `libzfs_core`.  Then this code could remain private, just in libzfs_core instead of libzfs.

Reply:
There; was that change (35bc51c) something like what you had in mind?

Reply:
Yup, that's what I had in mind.  Let just make sure to remove `libzfs_set_pipe_max` from `libzfs_impl.h` and make sure the replacement `lzc_*_pipe_max` functions are `static`.  We should be able to keep those functions private.